docs/root/intro/arch_overview/data_sharing_between_filters.rst

```diff
@@ -0,0 +1,94 @@
+HTTP routes use the metadata to indicate the labels on the endpoints
+corresponding to the weighted cluster. Another example, the subset load
+balancer uses the metadata from the route entry corresponding to the
+weighted cluster to select appropriate endpoints in a cluster
+
+Typed Metadata
+--------------
+
+:ref:`Metadata <envoy_api_msg_core.Metadata>` as such is untyped. Before
+acting on the metadata, callers typically convert it to a typed class
+object. The cost of conversion becomes non-negligible when performed
+repeatedly (e.g., for each request stream or connection). Typed Metadata
+solves this problem by allowing filters to register a one time conversion
+logic for a specific key. Incoming config metadata (via xDS) is converted
+to class objects at config load time. Filters can then obtain a typed
+variant of the metadata at runtime (per request or connection), thereby
+eliminating the need for filters to repeatedly convert from
+`ProtobufWkt::Struct` to some internal object during request/connection
+processing.
+
+For example, a filter that desires to have a convenience wrapper class over
+an opaque metadata with key `xxx.service.policy` in `ClusterInfo` could
+register a factory `ServicePolicyFactory` that inherits from
+`ClusterTypedMetadataFactory`. The factory translates the `ProtobufWkt::Struct`
+into an instance of `ServicePolicy` class (inherited from
+`FilterState::Object`). When a `Cluster` is created, the associated
+`ServicePolicy` instance will be created and cached. Note that typed
+metadata is not a new source of metadata. It is obtained from metadata that
+is specified as part of the configuration.
+
+
+HTTP Per-Route Filter Configuration
+-----------------------------------
+
+In HTTP routes, :ref:`per_filter_config
+<envoy_api_field_route.VirtualHost.per_filter_config>` is a special case of
+typed metadata where HTTP filters can have virtualhost/route-specific
+configuration in addition to a global filter config common to all virtual
+hosts. This configuration is converted and embedded into the route
+table. Its up to the HTTP filter implementation to treat the route-specific
+filter config as a replacement to global config or an enhancement. The HTTP
+fault filter uses this technique to provide per-route fault configuration.
+
+`per_filter_config` is a `map<string, ProtobufWkt::Struct>`. The Connection
+manager iterates over this map and invokes the filter factory interface
+`createRouteSpecificFilterConfig` to parse/validate the struct value and
+convert it into a typed class object that’s stored with the route
+itself. HTTP filters can then query the route-specific filter config during
+request processing.
+
+`per_filter_config` shares same traits and design goals as the typed
+metadata. Its a specialized implementation focussing on per-route filter
+configuration (whether the filter implementation treats it as metadata or
+actual configuration is upto the filter author).
```

Comment:
Not sure if this explanation actually makes it clearer in the end to the average filter writer (vs. to an audience consisting of us).